python/tvm/relay/quantize/quantize.py

```diff
@@ -0,0 +1,235 @@
+
+
+def current_qconfig():
+    """Get the current quantization configuration."""
+    return _quantize._GetCurrentQConfig()
+
+
+def qconfig(**kwargs):
+    """Configure the quantization behavior by setting config variables.
+
+    Parameters
+    ---------
+    nbit_dict: dict of QAnnotateKind -> int
+        Number of bit for every kind of annotate field.
+
+    global_scale: float
+        The global scale for calibration.
+
+    skip_k_conv: int
+        The number of skipped conv2d.
+
+    round_for_shift: boolean
+        Whether to add bias for rounding during shift.
+
+    store_lowbit_output: boolean
+        Whether to store low-bit integer back as output before dequantizing.
+        Some accelerators need this, e.g. VTA.
+
+    Returns
+    -------
+    config: QConfig
+        The quantization configuration
+    """
+    node_args = {k: v if k not in kwargs else kwargs[k]
+                 for k, v in QConfig._node_defaults.items()}
+    config = _make.node("relay.quantize.QConfig", **node_args)
+    return config
+
+
+CONV_COUNTER = 0
+
+
+def _conv_counter():
+    """Get the global counter for conv2d."""
+    return CONV_COUNTER
+
+
+def _set_conv_counter(n):
+    """Set the value of the global conv2d counter."""
+    global CONV_COUNTER
+    CONV_COUNTER = n
+
+
+def annotate(graph):
+    """Given a float32 graph, annotate will rewrite the graph
+    and return back a graph which simulates the error brought by
+    current quantization scheme.
+
+    Parameters
+    ---------
+    graph: Function
+        The original graph
+    """
+    _set_conv_counter(0)  # reset counter
+    return _quantize.annotate(graph)
+
+
+def calibrate(graph, dataset=None):
+    """The calibrate procedure will try to calculate the content of
+    dom_scale, nbit, clip_min, clip_max for every `simulated_quantize`
+    operator.
+
+    Parameters
+    ---------
+    graph: Function
+        The simulation graph after annotation.
+
+    dataset: list of dict of Var -> NDArray
+        The calibration dataset.
+    """
+    def power2_scale(arr):
+        """calculate weight scale with nearest mode-2 scale"""
+        val = np.amax(np.abs(arr.asnumpy()))
+        if val == 0.0:
+            return 1.0
+        else:
+            return 2**math.ceil(math.log(val, 2))
+
+    cfg = current_qconfig()
+    const_params = {}
+    quantize_op = _op.get("simulated_quantize")
+
+    def visit_func(e):
+        if isinstance(e, _expr.Call) and e.op == quantize_op:
+            ndata, ndom_scale, nclip_min, nclip_max = e.args
+            attrs = e.attrs
+            kind = attrs.kind
+            TABLE = [0, cfg.nbit_input, cfg.nbit_weight, cfg.nbit_activation]
+            nbit = TABLE[kind]
+
+            valid_bit = nbit
+            if attrs.sign:
+                valid_bit = nbit - 1
+
+            if kind == QAnnotateKind.WEIGHT:
+                var = e.args[0]
+                assert isinstance(var, _expr.Constant)
```

Comment:
This assertion fails if `canonicalize_ops` pass is enabled 
@merrymercy I guess this is intended that `BiasAddSimplifier` leaves constant unfoled?